compiler/noirc_evaluator/src/acir/arrays.rs

```diff
@@ -642,113 +601,99 @@ impl Context<'_> {
-                                            element_type_sizes,
-                                            type_sizes_array_len,
-                                        )?;
-                                        self.internal_mem_block_lengths
-                                            .insert(element_type_sizes, type_sizes_array_len);
-                                        return Ok(element_type_sizes);
-                                    } else {
-                                        return Err(InternalError::General {
-                                            message: format!("Array {array_id}'s inner element type sizes array should be initialized"),
-                                            call_stack: self.acir_context.get_call_stack(),
-                                        }
-                                        .into());
-                                    }
-                                }
-                            }
-                            AcirValue::Array(values) => {
-                                for (i, value) in values.iter().enumerate() {
-                                    flat_elem_type_sizes.push(
-                                        Self::flattened_value_size(value) + flat_elem_type_sizes[i],
-                                    );
-                                }
-                            }
-                            _ => {
-                                return Err(InternalError::Unexpected {
-                                    expected: "AcirValue::DynamicArray or AcirValue::Array"
-                                        .to_owned(),
-                                    found: format!("{:?}", array_acir_value),
-                                    call_stack: self.acir_context.get_call_stack(),
-                                }
-                                .into());
-                            }
-                        }
-                    }
-                    _ => {
-                        return Err(InternalError::Unexpected {
-                            expected: "array or instruction".to_owned(),
-                            found: format!("{:?}", &dfg[array_id]),
-                            call_stack: self.acir_context.get_call_stack(),
-                        }
-                        .into());
-                    }
-                };
+        if !matches!(array_typ, Type::Array(_, _) | Type::Slice(_)) {
+            return Err(InternalError::Unexpected {
+                expected: "array or slice".to_owned(),
+                found: array_typ.to_string(),
+                call_stack: self.acir_context.get_call_stack(),
             }
-            _ => {
-                return Err(InternalError::Unexpected {
-                    expected: "array or slice".to_owned(),
-                    found: array_typ.to_string(),
-                    call_stack: self.acir_context.get_call_stack(),
-                }
-                .into());
```

Comment:
I've replaced the many levels of nesting with some early returns. The code which originally sat below the `match` statement now sits in the `AcirValue::Array` arm.